Update documentation on experiments and A/B tests

shadyvb opened the pull request:
This PR reorganizes the docs on personalization and experimentation features of Altis Analytics, to encapsulate both under the term 'Optimization framework', as well as adding docs on the new framework for registering Post A/B tests after updating the framework to be easily extensible.

The updates are based on a discussion that happened recently, trying to make sense of the hierarchical nature of the framework we have, so the docs are a bit more clear and understandable for visitors.

https://github.com/humanmade/product-dev/issues/858

shadyvb commented:
@roborourke to see the changes a bit more clearly, try previewing them locally by setting the `mainBranch` of this repo within the docs site code to `docs-optimization-framework` and building/previewing the output.

roborourke commented on docs/optimization-framework/ab-testing.md:
```diff
@@ -1,19 +1,14 @@
+# A/B Tests
 
-Backed by [native Altis Analytics](./native.md), Experiments are a powerful tool for optimizing content and measuring the effectiveness of changes to the site.
+A/B Tests provides the tools to carry out randomized experiments using different variations of content features, eg: titles and featured images, and/or different variations of content blocks.
```
Not sure on the use of the word randomized just here, to me it read like the experiment was random, rather than the assignment of users to different variants for example.
```suggestion
A/B Tests provides the tools to carry out experiments using different variations of content features, eg: titles and featured images, and/or different variations of content blocks.
```

roborourke commented on docs/optimization-framework/ab-testing.md:
```diff
@@ -1,19 +1,14 @@
+A/B Tests provides the tools to carry out randomized experiments using different variations of content features, eg: titles and featured images, and/or different variations of content blocks.
 
-The feature is enabled by default and provides a developer API for creating custom experiments as well as some built in features.
+Altis provides an A/B Test Block, which you can use to test different variations of a block / set of blocks. And also a developer friendly framework to test different content features like Title and Featured Images, which can typically be controlled from the Block Editor sidebar.
```
Should never really start a sentence with "And", plus I think we can add a bit of detail e.g.:
```suggestion
Altis provides an A/B Test Block, which you can use to test different variations of a block or set of blocks in your content, widgets and layouts. There is also a developer friendly framework to add new custom tests that is used to power built in features like Title and Featured Image tests which can be controlled from the Block Editor sidebar.
```

roborourke commented on docs/optimization-framework/ab-testing.md:
````diff
@@ -35,20 +31,24 @@ It is enabled by default but can be disabled via the configuration file:
 ```
 
-Within the post edit screen click on the A/B icon to access the Experiments panel. Here you can set the alternative titles, the amount of traffic to show the tests to as well as start and end dates.
+Within the post edit screen click on the A/B icon to access the Experiments panel. Here you can set the alternative titles and/or alternative featured images, the amount of traffic to show the tests to as well as start and end dates.
````
Think we go a bit more generic perhaps to avoid having to edit lots of it in future.
```suggestion
Within the post edit screen click on the A/B icon to access the Experiments panel. Here you can set up the variants for your tests, the amount of traffic to show the tests to as well as start and end dates.
```

roborourke commented on docs/optimization-framework/ab-testing.md:
````diff
@@ -35,20 +31,24 @@ It is enabled by default but can be disabled via the configuration file:
 ![A/B Testing Titles user interface](../assets/ab-tests-titles.png)
 
-By default post title A/B tests are enabled for Posts and Pages however custom post types can be supported using the `altis.experiments.titles` key either when registering the post type or in the `init` action, for example:
+By default post title and featured image A/B tests are enabled for Posts and Pages however custom post types can be supported using the `altis.experiments.titles` key either when registering the post type or in the `init` action, for example:
 
 ```php
 add_action( 'init', function () {
 	add_post_type_support( 'events', 'altis.experiments.titles' );
+	// and/or..
+	add_post_type_support( 'events', 'altis.experiments.featured_images' );
````
Minor nitpick but I try to avoid using a slash like this in the public facing docs, plus "and/or" is saying "and or or"
```suggestion
By default post title and featured image A/B tests are enabled for Posts and Pages however custom post types can be supported using keys shown in the example below, either when registering the post type or on the `init` action, for example:

```php
add_action( 'init', function () {
	// Add support for title A/B tests to the 'events' post type.
	add_post_type_support( 'events', 'altis.experiments.titles' );

	// Add support for featured image A/B tests to the 'events' post type.
	add_post_type_support( 'events', 'altis.experiments.featured_images' );
```
```

roborourke commented on docs/optimization-framework/ab-testing.md:
````diff
@@ -35,20 +31,24 @@ It is enabled by default but can be disabled via the configuration file:
 } );
 ```
 
+Note that in order for the test to function correctly, it needs WordPress to output the `<ab-test></ab-test>` Web Component that includes all the variations of the feature representation, so the front-end would be able to switch to the proper variation to serve to the visitor. And such, the built-in features will only work if the native functions are used, which are `the_title` / `get_the_title`  ...
````
```suggestion
Note that in order for the test to function correctly, it needs WordPress to output the `<ab-test></ab-test>` Web Component (generated by the `output_ab_test_html_for_post()` function) that includes all the variations of the feature representation, so the front-end would be able to switch to the proper variation to serve to the visitor. As such, the built-in features will only work if the native WordPress functions are used, which are `the_title()` or `get_the_title()` for post title and `the_post_thumbnail()` or `get_the_post_thumbnail()` for featured image.
```

roborourke replied:
This also makes me think we should document the web components too like we have here currently for personalization https://docs.altis-dxp.com/v8/analytics/native/experience-blocks/#personalization

So something like "it needs WordPress to output the `<ab-test></ab-test>` or `<ab-test-block></ab-test-block>` Web Component". Then perhaps to link to a separate sub page or just document them inline.

For reference `<ab-test>` is designed for inline content and has no impact on the layout or styling of the site. `<ab-test-block>` is for block level content and will output the content in a wrapper with the default CSS `display: block`.

shadyvb replied:
I had started to document the `<ab-test/>` component with a headless use case, but stopped mid-way realizing that I don't really know how it works yet 😅 and that there is quite a bit to document rather than just the block. So I'd probably exclude that as a goal from the update here. A new issue would be a bit more efficient.

roborourke commented on docs/optimization-framework/ab-testing.md:
```diff
@@ -85,30 +87,141 @@ Returns the A/B Test markup for client side processing.
+				'wp-components',
+				'wp-core-data',
+				'wp-edit-post',
+				'...',
```
I'd just leave this out, ideally it should be more or less directly copy/pastable code.

shadyvb replied:
I'll keep one entry only for the sake of demoing how it can be used.

roborourke commented on docs/optimization-framework/ab-testing.md:
````diff
@@ -85,30 +87,141 @@ Returns the A/B Test markup for client side processing.
+} );
+```
+
+To register a custom editor input, and override some of the test options.
````
```suggestion
To register a custom editor input, and override some of the test options:
```

roborourke commented on docs/optimization-framework/experience-blocks.md:
```diff
@@ -2,11 +2,11 @@
+## Personalization API
 
-Altis empowers editorial teams to create personalized content with the Personalized Content Experience Block. Because it's crucial that web pages can be fully cached for performance, our Personalized Content blocks use [web components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) that are simple to use and load immediately as the document is loaded.
+Altis empowers editorial teams to create personalized content with the Personalized Content Experience Block. And because it's crucial that web pages can be fully cached for performance, our Personalized Content blocks use [web components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) that are simple to use and load immediately as the document is loaded.
```
Remember, don't start a sentence with "And", either switch to a comma or leave as-is. I see how it could perhaps read as a direct follow on from the first sentence but the full stop was intentional. How about this small tweak:

```suggestion
Altis empowers editorial teams to create personalized content with the Personalized Content Experience Block. Because it is crucial that web pages can be fully cached for performance, our Personalized Content blocks use [web components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) that are simple to use and load immediately as the document is loaded.
```

roborourke commented on docs/optimization-framework/ab-testing.md:
````diff
@@ -85,30 +87,141 @@ Returns the A/B Test markup for client side processing.
+```js
+// feature.js
+
+import { EditControl, PreviewControl } from 'some/custom/library';
````
I don't really understand this JS example and I'm not confident that I could use it to create my own custom test as it is. Can we use a more complete real-world example? Maybe take a pared down version of the title test implementation that changes the output of some post meta or custom field?

roborourke commented on docs/optimization-framework/ab-testing.md:
```diff
@@ -85,30 +87,141 @@ Returns the A/B Test markup for client side processing.
 	] );
-}, 10, 5 );
+
+	// Apply the test by using `output_ab_test_html_for_post`.
+	add_action( 'the_custom_cta', function () {
+		$post_id = get_the_ID();
+
+		return output_ab_test_html_for_post( 'cta_button_text', $post_id, $some_extra_args );
+	}, 10, 5 );
```
This could be added outside of the `init` action, same in the example below.

roborourke commented on docs/optimization-framework/ab-testing.md:
```diff
@@ -85,30 +87,141 @@ Returns the A/B Test markup for client side processing.
 - `$default_content`: The default content for the control test.
 - `$args`: An optional array of data to pass through to `variant_callback`.
 
+### Example basic test
```
```suggestion
### Custom Test Example
```

roborourke commented on docs/optimization-framework/ab-testing.md:
```diff
@@ -85,30 +87,141 @@ Returns the A/B Test markup for client side processing.
 
+### Example basic test
+
+An example for testing a CTA (call to action) button with different text would look like:
```
It's hard from the code example to see how this would be written or used from start to finish. I think it needs to be something easier to grok and more real-world so perhaps we can reuse the bulk of the code for the title test but apply it to a piece of post meta data, then follow that through showing the post meta being filtered and used in a template part on output.

roborourke commented on docs/native/api/goal-tracking.md:
```diff
@@ -1,7 +1,7 @@
 # Goal Tracking
 
-Conversions goals help you measure the performance of aspects of your site. With Altis Native Analytics you can use the [client-side API](./client-side-api.md) to track any events you wish, however, some goal tracking is integrated into the [Experiments](./experiments.md) and [Experience Blocks](./experience-blocks.md) features.
+Conversions goals help you measure the performance of aspects of your site. With Altis Native Analytics you can use the [client-side API](./client-side-api.md) to track any events you wish, however, some goal tracking is integrated into [Altis Optimization Framework](../../optimization-framework/) features like [A/B Testing](../../optimization-framework/ab-testing.md) and [Experience Blocks](../. ...
```
I think this goal tracking document should live under the Optimization Framework section to be honest as it's really tied to that. The native analytics portion itself doesn't really do anything with it by default.

shadyvb replied:
I had updated the link in the `ab-testing.md` but didn't move the file, good catch!

roborourke reviewed (commented):
I like the reorganisation 😄 Definitely a massive improvement so far.

Suggested some tweaks and I think it still needs a bit more work, most importantly:

- Move the goal tracking doc under the optimization framework section
- Make the custom A/B test example a complete real-world example, e.g. for setting variants of a custom meta field used to display share button text, I find it a bit too abstract to understand at the moment


Less important but nice to have would be to briefly document the `<ab-test>` and `<ab-test-block>` web components for direct use like we have with the `<personalization-block>`

shadyvb commented:
I had started to document the `<ab-test/>` component with a headless use case, but stopped mid-way realizing that I don't really know how it works yet 😅 and that there is quite a bit to document rather than just the block. So I'd probably exclude that as a goal from the update here. A new issue would be a bit more efficient.

roborourke commented on docs/optimization-framework/ab-testing.md:
```diff
@@ -0,0 +1,311 @@
+
+- `$test_id`: A unique ID for the test.
+- `$options`: Configuration options for the test.
+  - `label <string>`: A human readable label for the test.
```
Did we add another item here? `singular_label`? Should probably add those to example further down too

shadyvb replied:
Yeah, I even missed adding `label` to the examples, will do once we rewrite them as asked. Added to the function doc for now.

shadyvb commented:
@roborourke I addressed the feedback, split the custom test example into its own file, and provided more complete steps to go about it.
Also updated the notes on featured image test to reflect how we're not filtering `has_post_thumbnail` for logical reasons.

Back to your review.

roborourke commented on docs/optimization-framework/custom-ab-test-example.md:
````diff
@@ -0,0 +1,125 @@
+Step 3: Output the test output in your template.
+
+```php
+// single-event.php
+
+// ..
+<div class='cta-container'>
+	<?php output_ab_test_html_for_post( 'cta_button_style', get_the_ID() ); ?>
+</div>
+// ..
````
This function returns a string so will need to be echoed, maybe can pass the fallback output parameter too?

Also thinking this file could be a template part, like `page-cta.php`, so rather than truncating it with `// ..` it can be a complete file including namespacing:

```suggestion
<?php
// template-parts/cta.php

use Altis\Analytics\Experiments;
 
$default = sprintf(
	'<button class="cta-button cta-button-medium">%s</button>',
	__( 'Register Now', 'my-theme' )
);

?>
<div class="cta-container">
	<?php
		echo Experiments\output_ab_test_html_for_post(
			'cta_button_style',
			get_the_ID(),
			$default
		);
	?>
</div>
```

roborourke commented on docs/optimization-framework/custom-ab-test-example.md:
```diff
@@ -0,0 +1,125 @@
+		'goal' => 'click',
+		'variant_callback' => function ( $variant_value, $post_id ) {
+			$label = __( 'Register now', 'my-theme' );
+			return `
+				<button class="cta-button cta-button-$variant_value">
+					$label
+				</button>
+			`;
```
Backticks in PHP are used to execute code as a shell command, so:
```suggestion
			return sprintf(
				'<button class="cta-button cta-button-%s">%s</button>',
				$variant_value,
				$label
			);
```

roborourke commented on docs/optimization-framework/custom-ab-test-example.md:
````diff
@@ -0,0 +1,125 @@
+Step 1: Register the test, and the custom meta field that will control the value we'll be testing.
+
+```php
+// functions.php
````
```suggestion
<?php
// functions.php
```

roborourke reviewed (approved):
Awesome, couple of minor suggestions on the code examples for completeness and we're good. Nice one @shadyvb !
